File: 2026-08/src/day_06/leet_49/Solution.java
package day_06.leet_49;

import java.util.*;

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> ans = new ArrayList<>();

        out:
        for (String str : strs) {
            for (List<String> list : ans) {
                if (list.get(0).length() == str.length() && isAnagram(list.get(0), str)) {
                    list.add(str);
                    continue out;
                }
            }

            List<String> list = new ArrayList<>();
            list.add(str);
            ans.add(list);
        }

        return ans;
    }

    static boolean isAnagram(String s1, String s2) {
        int[] cnt = new int[26];
        for (char c : s1.toCharArray()) {
            cnt[c - 'a']++;
        }
        for (char c : s2.toCharArray()) {
            cnt[c - 'a']--;
        }

        for (int x : cnt) {
            if (x != 0) return false;
        }

        return true;
    }
}

File: 2026-08/src/day_06/leet_49/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
   public:
    bool isAnagram(string& s1, string& s2) {
        vector<int> cnt(26);
        for (char c : s1) {
            cnt[c - 'a']++;
        }
        for (char c : s2) {
            cnt[c - 'a']--;
        }

        for (int x : cnt) {
            if (x != 0) return false;
        }

        return true;
    }

    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> ans;

        for (string& str : strs) {
            bool flag = false;

            for (vector<string>& v : ans) {
                if (v[0].size() == str.size() && isAnagram(v[0], str)) {
                    v.push_back(str);
                    flag = true;
                    break;
                }
            }

            if (flag) continue;

            vector<string> v;
            v.push_back(str);
            ans.push_back(v);
        }

        return ans;
    }
};
